tutorials/adult_tutorial.py

```diff
@@ -0,0 +1,178 @@
+            opt_loss = scalar_loss
+        global_step = tf.compat.v1.train.get_global_step()
+        train_op = optimizer.minimize(loss=opt_loss, global_step=global_step)
+        # In the following, we pass the mean of the loss (scalar_loss) rather than
+        # the vector_loss because tf.estimator requires a scalar loss. This is only
+        # used for evaluation and debugging by tf.estimator. The actual loss being
+        # minimized is opt_loss defined above and passed to optimizer.minimize().
+        return tf.estimator.EstimatorSpec(mode=mode,
+                                          loss=scalar_loss,
+                                          train_op=train_op)
+
+    # Add evaluation metrics (for EVAL mode).
+    if mode == tf.estimator.ModeKeys.EVAL:
+        eval_metric_ops = {
+            'accuracy':
+                tf.compat.v1.metrics.accuracy(
+                    labels=labels,
+                    predictions=tf.argmax(input=logits, axis=1))
+        }
+        return tf.estimator.EstimatorSpec(mode=mode,
+                                          loss=scalar_loss,
+                                          eval_metric_ops=eval_metric_ops)
+
+    return None
+
+
+def load_adult():
+    """Loads ADULT a2a as in LIBSVM and preprocesses to combine training and validation data."""
+    # https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html
+
+    X = pd.read_csv("adult.csv")
+    kf = KFold(n_splits=10)
+    for train_index, test_index in kf.split(X):
+        train, test = X.iloc[train_index, :], X.iloc[test_index, :]
+    train_data = train.iloc[:, range(X.shape[1]-1)].values.astype('float32')
+    test_data = test.iloc[:, range(X.shape[1]-1)].values.astype('float32')
+
+    train_labels = (train.iloc[:, X.shape[1]-1] == 1).astype('int32').values
+    test_labels = (test.iloc[:, X.shape[1]-1] == 1).astype('int32').values
+
+    return train_data, train_labels, test_data, test_labels
+
+
+def main(unused_argv):
+    '''main'''
+    tf.compat.v1.logging.set_verbosity(0)
+
+    # Load training and test data.
+    train_data, train_labels, test_data, test_labels = load_adult()
+
+    # Instantiate the tf.Estimator.
+    adult_classifier = tf.compat.v1.estimator.Estimator(model_fn=nn_model_fn,
+                                                        model_dir=FLAGS.model_dir)
+
+    # Create tf.Estimator input functions for the training and test data.
+    eval_input_fn = tf.compat.v1.estimator.inputs.numpy_input_fn(
+        x={'x': test_data},
+        y=test_labels,
+        num_epochs=1,
+        shuffle=False)
+
+    # Training loop.
+    steps_per_epoch = 29305 // 256
+    test_accuracy_list = []
+    for epoch in range(1, FLAGS.epochs + 1):
+        for step in range(steps_per_epoch):
+            whether = np.random.random_sample(29305) > (1-256/29305)
+            subsampling = [i for i in np.arange(29305) if whether[i]]
+            global microbatches
+            microbatches = len(subsampling)

```

Comment:
this is confusing. Why do you need to set the global microbatches? Probably there is a safer way to do this.